File: pkg/transport/udp.go
package transport

import (
	"errors"
	"net"
	"sync"
	"time"

	log "github.com/sirupsen/logrus"
)

// maxDatagram is the largest datagram we are willing to read. Anything larger
// cannot be sent over UDP anyway.
const maxDatagram = 65535

// dialTimeout bounds a stream dial so an unreachable peer cannot stall the
// caller (state sync runs on a background goroutine, but a joiner waits on it).
const dialTimeout = 5 * time.Second

// UDPTransport sends and receives datagrams on a single shared socket, and
// accepts reliable streams on the same address over TCP.
//
// One socket for both directions matters: opening a fresh socket per packet
// (as this code used to) burns a file descriptor and an ephemeral port per
// message, and makes every packet appear to come from a different port.
type UDPTransport struct {
	conn     *net.UDPConn
	listener net.Listener
	addr     string

	packets chan Packet
	streams chan net.Conn

	resolveMu sync.Mutex
	resolved  map[string]*net.UDPAddr

	closeOnce sync.Once
	closed    chan struct{}
}

// ListenUDP binds addr for datagrams and, when possible, the same address for
// streams. A failure to bind the stream listener is not fatal: the node still
// works, falling back to datagram state sync.
func ListenUDP(addr string) (*UDPTransport, error) {
	udpAddr, err := net.ResolveUDPAddr("udp", addr)
	if err != nil {
		return nil, err
	}
	conn, err := net.ListenUDP("udp", udpAddr)
	if err != nil {
		return nil, err
	}
	t := &UDPTransport{
		conn:     conn,
		addr:     addr,
		packets:  make(chan Packet, 256),
		streams:  make(chan net.Conn, 16),
		resolved: make(map[string]*net.UDPAddr),
		closed:   make(chan struct{}),
	}
	if ln, err := net.Listen("tcp", addr); err != nil {
		log.Errorf("stream listener on %s unavailable, falling back to datagram state sync: %s", addr, err)
	} else {
		t.listener = ln
		go t.acceptLoop()
	}
	go t.readLoop()
	return t, nil
}

func (t *UDPTransport) LocalAddr() string { return t.addr }

func (t *UDPTransport) Packets() <-chan Packet { return t.packets }

func (t *UDPTransport) Streams() <-chan net.Conn { return t.streams }

func (t *UDPTransport) Send(addr string, data []byte) error {
	udpAddr, err := t.resolve(addr)
	if err != nil {
		return err
	}
	_, err = t.conn.WriteToUDP(data, udpAddr)
	return err
}

func (t *UDPTransport) Dial(addr string) (net.Conn, error) {
	if t.listener == nil {
		return nil, errors.New("stream transport unavailable")
	}
	return net.DialTimeout("tcp", addr, dialTimeout)
}

func (t *UDPTransport) Close() error {
	var err error
	t.closeOnce.Do(func() {
		close(t.closed)
		if t.listener != nil {
			t.listener.Close()
		}
		err = t.conn.Close()
	})
	return err
}

// resolve caches address lookups: peer addresses repeat on every gossip tick,
// and resolving each time turns a hostname peer into a DNS query per packet.
func (t *UDPTransport) resolve(addr string) (*net.UDPAddr, error) {
	t.resolveMu.Lock()
	defer t.resolveMu.Unlock()
	if a, ok := t.resolved[addr]; ok {
		return a, nil
	}
	a, err := net.ResolveUDPAddr("udp", addr)
	if err != nil {
		return nil, err
	}
	t.resolved[addr] = a
	return a, nil
}

func (t *UDPTransport) readLoop() {
	defer close(t.packets)
	buf := make([]byte, maxDatagram)
	for {
		n, from, err := t.conn.ReadFromUDP(buf)
		if err != nil {
			if errors.Is(err, net.ErrClosed) {
				return
			}
			log.Errorf("read udp: %s", err)
			continue
		}
		data := make([]byte, n)
		copy(data, buf[:n])
		pkt := Packet{From: from.String(), Data: data}
		select {
		case t.packets <- pkt:
		case <-t.closed:
			return
		}
	}
}

func (t *UDPTransport) acceptLoop() {
	defer close(t.streams)
	for {
		conn, err := t.listener.Accept()
		if err != nil {
			if errors.Is(err, net.ErrClosed) {
				return
			}
			log.Errorf("accept stream: %s", err)
			continue
		}
		select {
		case t.streams <- conn:
		case <-t.closed:
			conn.Close()
			return
		}
	}
}

File: electron/src/net/udp.js
'use strict';

const dgram = require('node:dgram');
const net = require('node:net');
const { EventEmitter } = require('node:events');
const { targetOf } = require('./addr');

/** Bounds a stream dial so an unreachable peer cannot stall a joiner. */
const DIAL_TIMEOUT_MS = 5_000;
/** Bounds one request/response exchange on a stream. */
const STREAM_TIMEOUT_MS = 10_000;

/**
 * Datagrams and streams on one port.
 *
 * Datagrams carry the protocol; streams carry payloads that do not fit one — a
 * full state sync or a large bootstrap roster. Both are framed identically, the
 * stream form simply adding a length prefix.
 *
 * One socket for everything matters: a fresh socket per packet burns a
 * descriptor and an ephemeral port per message, and makes every packet appear
 * to come from a different port.
 */
class UdpTransport extends EventEmitter {
  constructor(socket, server, port) {
    super();
    this.socket = socket;
    this.server = server;
    this.port = port;
    this.closed = false;
    socket.on('message', (data, rinfo) => {
      this.emit('packet', { from: `${rinfo.address}:${rinfo.port}`, data: Buffer.from(data) });
    });
    socket.on('error', (e) => {
      if (!this.closed) this.emit('error', { what: 'datagram socket', error: e });
    });
    if (server) {
      server.on('connection', (conn) => {
        conn.setTimeout(STREAM_TIMEOUT_MS, () => conn.destroy());
        conn.on('error', () => {});
        this.emit('stream', conn);
      });
      server.on('error', (e) => {
        if (!this.closed) this.emit('error', { what: 'stream listener', error: e });
      });
    }
  }

  /**
   * Binds port for datagrams and, when possible, the same port for streams. A
   * failure to bind the stream listener is not fatal: the node still works,
   * falling back to datagram state sync.
   */
  static async listen(port, opts = {}) {
    const host = opts.host || '0.0.0.0';
    const socket = dgram.createSocket({ type: 'udp4', reuseAddr: true });
    await new Promise((resolve, reject) => {
      const onError = (e) => {
        socket.removeListener('listening', onListening);
        reject(e);
      };
      const onListening = () => {
        socket.removeListener('error', onError);
        resolve();
      };
      socket.once('error', onError);
      socket.once('listening', onListening);
      socket.bind(port, host);
    });
    const bound = socket.address().port;

    let server = null;
    try {
      server = await new Promise((resolve, reject) => {
        const srv = net.createServer();
        srv.once('error', reject);
        srv.listen(bound, host, () => {
          srv.removeListener('error', reject);
          resolve(srv);
        });
      });
    } catch (e) {
      server = null;
      if (opts.onError) opts.onError('stream listener unavailable', e);
    }
    return new UdpTransport(socket, server, bound);
  }

  /** False when the TCP port was taken, which drops state sync back to datagrams. */
  get streamsAvailable() {
    return this.server !== null;
  }

  /** Hands one datagram to the kernel. Resolves when handed over, never when delivered. */
  send(addr, data) {
    const target = targetOf(addr);
    if (!target) return Promise.reject(new Error(`unusable address: ${addr}`));
    if (this.closed) return Promise.reject(new Error('transport closed'));
    return new Promise((resolve, reject) => {
      this.socket.send(data, 0, data.length, target.port, target.host, (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  /** Opens a reliable stream to addr, or rejects when the stream path is unavailable. */
  dial(addr) {
    const target = targetOf(addr);
    if (!target) return Promise.reject(new Error(`unusable address: ${addr}`));
    if (!this.server) return Promise.reject(new Error('stream transport unavailable'));
    return new Promise((resolve, reject) => {
      const conn = net.connect({ host: target.host, port: target.port });
      const timer = setTimeout(() => {
        conn.destroy();
        reject(new Error(`dial ${addr} timed out`));
      }, DIAL_TIMEOUT_MS);
      conn.once('connect', () => {
        clearTimeout(timer);
        conn.setTimeout(STREAM_TIMEOUT_MS, () => conn.destroy());
        resolve(conn);
      });
      conn.once('error', (e) => {
        clearTimeout(timer);
        reject(e);
      });
    });
  }

  close() {
    if (this.closed) return Promise.resolve();
    this.closed = true;
    const closing = [];
    closing.push(new Promise((resolve) => {
      try {
        this.socket.close(() => resolve());
      } catch {
        resolve();
      }
    }));
    if (this.server) {
      closing.push(new Promise((resolve) => this.server.close(() => resolve())));
    }
    return Promise.all(closing).then(() => undefined);
  }
}

module.exports = { UdpTransport, DIAL_TIMEOUT_MS, STREAM_TIMEOUT_MS };
